src/server/frontend_wayland/generator/enum.cpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright © 2018 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored By: William Wold <william.wold@canonical.com>
+ */
+
+#include "enum.h"
+#include "utils.h"
+
+#include <libxml++/libxml++.h>
+
+Enum::Enum(xmlpp::Element const& node)
+    : name{sanitize_name(to_camel_case(node.get_attribute_value("name")))}
+{
+    for (auto const& child : node.get_children("entry"))
+    {
+        auto entry_node = dynamic_cast<xmlpp::Element const*>(child);
+        std::string entry_name = sanitize_name(entry_node->get_attribute_value("name"));
+        std::string entry_value_str = entry_node->get_attribute_value("value");
+        entries.emplace_back(Entry{entry_name, entry_value_str});
+    }
+}
+
+Emitter Enum::declaration() const
+{
+    std::vector<Emitter> body;
+    for (auto const& entry : entries)
+    {
+        body.push_back({"static uint32_t const ", entry.name, " = ", entry.value, ";"});
+    }
+
+    return Lines{
+        {"struct ", name},
```

Comment:
Hm. Why are we translating a Wayland enum into a `struct` with some static members? Wouldn't this be more naturally translated to an `enum class : uint32_t` (for enums without `bitfield=true` set). For `bitfield=true` Wayland enums, you *could* try something like [this](https://www.justsoftwaresolutions.co.uk/cplusplus/using-enum-classes-as-bitfields.html)?

I think that we should at least translate non-bitfields into `enum class`, rather then `struct` + static constants.

Reply:
That bitfield enum class thing looks complicated (but interesting!). I didn't want to deal with something like that now, or force arguments to know what enum they are, so did it the easy way. I agree this would be beneficial, but it is an enhancement.

Reply:
I'm actually not sure enum class is the right choice for bit fields. When you `or` them, you just get the underlying value type, not the enum type, right? If so, we loose type safety which defeats the point. Any one of those statements Could be incorrect, that's just my first impression.

Reply:
Perhaps we could send around initializer lists of enum classes... I'll look into it.